pageserver/src/layered_repository/storage_layer.rs

```diff
@@ -0,0 +1,113 @@
+use crate::relish::RelishTag;
+use crate::repository::WALRecord;
+use crate::walredo::WalRedoManager;
+use crate::ZTimelineId;
+use anyhow::Result;
+use bytes::Bytes;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+use zenith_utils::lsn::Lsn;
+
+pub static ZERO_PAGE: Bytes = Bytes::from_static(&[0u8; 8192]);
+
+///
+/// Represents a version of a page at a specific LSN. The LSN is the key of the
+/// entry in the 'page_versions' hash, it is not duplicated here.
+///
+/// A page version can be stored as a full page image, or as WAL record that needs
+/// to be applied over the previous page version to reconstruct this version.
+///
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PageVersion {
+    /// an 8kb page image
+    pub page_image: Option<Bytes>,
+    /// WAL record to get from previous page version to this one.
+    pub record: Option<WALRecord>,
+}
+
+///
+/// A Layer holds all page versions for one relish, in a range of LSNs.
+/// There are two kinds of layers, in-memory and snapshot layers. In-memory
+/// layers are used to ingest incoming WAL, and provide fast access
+/// to the recent page versions. Snaphot layers are stored on disk, and
+/// are immutable.
+///
+/// Each layer contains a full snapshot of the relish at the start
+/// LSN. In addition to that, it contains WAL (or more page images)
+/// needed to recontruct any page version up to the end LSN.
+///
+pub trait Layer: Send + Sync {
+    // These functions identify the relish and the LSN range that this Layer
+    // holds.
+    fn get_timeline_id(&self) -> ZTimelineId;
+    fn get_relish_tag(&self) -> RelishTag;
+    fn get_start_lsn(&self) -> Lsn;
+    fn get_end_lsn(&self) -> Lsn;
+    fn is_dropped(&self) -> bool;
+
+    /// Frozen layers are stored on disk, an cannot accept cannot accept new WAL
+    /// records, whereas an unfrozen layer can still be modified, but is not
+    /// durable in case of a crash. Snapshot layers are always frozen, and
+    /// in-memory layers are always unfrozen.
+    fn is_frozen(&self) -> bool;
```

Comment:
Right. Only one un-frozen top layer for each relish. This returns constant 'true' for InMemoryLayers and constant 'false' for SnapshotLayers. The checkpoint code deals with `dyn Layer`s but needs to somehow distinct between the two, that's what this function is for.

I think it would make sense to keep the InMemoryLayers and SnapshotLayers more separate in the LayerMap. Only the code that wants to read a page should treat them the same, most other code, like opening a layer for writing and checkpointing/freezing the layers, needs to distinguish them. I'll look into refactoring that, and I think this function can then go away too.